### DISCUSSION TITLE
希望增加“用mpv播放”的按钮

### DISCUSSION THREAD

**Discussion by Lee-7723** [the1812/Bilibili-Evolved#1762] [功能建议 / Ideas]
[MPV播放器系列(一)-剪辑在线视频-知乎](https://zhuanlan.zhihu.com/p/366296441)
刷知乎的时候刷到了这篇文章，提到
![image](https://user-images.githubusercontent.com/52600662/115545401-581e1680-a2d6-11eb-90d1-07dddbdbbe80.png)
mpv是我的主力播放器，本身自带了youtube-dl，但是直接用命令
`mpv "https://www.bilibili.com/video/BVxxxxxxxx(B站视频链接)"`
这样的方法的话，youtube-dl要先解析再播放，解析的时间有点长，而且youtube-dl本身我写了config文件，导致功能设置有点混乱，不太好用
**另一种方法是用B站视频直链，希望增加的也是这个功能**
B站的直链应该是加上Referer和User-Agent就能用了吧（我一萌新，说错了别打我），正好mpv支持指定Referer和User-Agent
![image](https://user-images.githubusercontent.com/52600662/115546160-46893e80-a2d7-11eb-835d-9987aba7af81.png)
[mpv的相关的操作说明](https://mpv.io/manual/master/#options-http-header-fields)

举例写法是：
`mpv --http-header-fields='referer:https://www.bilibili.com/','user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.72 Safari/537.36 Edg/90.0.818.39' "https://cn-tj2-cmcc-bcache-08.bilivideo.com/upgcxcode/03/08/327030803/327030803-1-116.flv?e=ig8euxZM2rNcNbRjhbdVhwdlhWTghwdVhoNvNC8BqJIzNbfqXBvEqxTEto8BTrNvN0GvT90W5JZMkX_YN0MvXg8gNEV4NC8xNEV4N03eN0B5tZlqNxTEto8BTrNvNeZVuJ10Kj_g2UB02J0mN0B5tZlqNCNEto8BTrNvNC7MTX502C8f2jmMQJ6mqF2fka1mqx6gqj0eN0B599M=&uipk=5&nbs=1&deadline=1618994931&gen=playurlv2&os=bcache&oi=3746057939&trid=981785f4d159442ea0ff8a6585885155u&platform=pc&upsig=c7a11cd17258281c6216e4ef07195f66&uparams=e,uipk,nbs,deadline,gen,os,oi,trid,platform&cdnid=6779&mid=310265481&orderid=0,3&agrr=1&logo=80000000"`

我试了一下，完美播放
![image](https://user-images.githubusercontent.com/52600662/115546620-d929dd80-a2d7-11eb-9194-1d36410dd15e.png)

**所以功能设想是**：
获取视频直链，然后生成一条cmd命令
`mpv --http-header-fields='referer:https://www.bilibili.com/','user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.72 Safari/537.36 Edg/90.0.818.39' "<direct video url>"`
一个按钮执行命令，这样把视频拉取到mpv看（当然前提是mpv在环境路径里）
这功能的用户可能会比较少，但是应该不难写吧（站着说话不腰疼😁），希望增加这样的一个按钮

**Comment by the1812** [comment]
网页不能直接跑cmd的, 你可以弄个批处理之类的 (类似 #1472)

**Reply by Lee-7723** [reply]
那bat内就是这样呗，不过要把\<direct url\>传递进来，我完全不懂脚本的运行方式，只是希望有个这样的功能
`mpv --http-header-fields="referer:https://www.bilibili.com/","user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.72 Safari/537.36 Edg/90.0.818.39" "<direct url>"`

**Reply by the1812** [reply]
`<direct url>` 应该可以靠剪贴板传递? 我不知道 bat 里能不能拿剪贴板
反正脚本是不能直接运行的, 顶多把这行命令文字生成出来, 最后还是要你复制粘贴到 cmd 里, 或者下载为 bat 再运行.
所以我建议是写个 bat, 脚本这边复制下链接, 你双击下那个 bat 就能播放了

**Reply by Lee-7723** [reply]
运行命令这个操作是我在玩VRChat的时候从网页上直接启动游戏，弹出的命令行窗口，我去再看一下

**Reply by Lee-7723** [reply]
![image](https://user-images.githubusercontent.com/52600662/115653645-1f2b8380-a362-11eb-9312-7b3d79c64d29.png)
VRChat打开是大概这个样子

[另外参考一下这个脚本](https://greasyfork.org/zh-CN/scripts/416271-play-with-mpv)

**Reply by the1812** [reply]
那是需要程序去[关联URL协议](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/platform-apis/aa767914(v=vs.85)?redirectedfrom=MSDN)的, 不知道mpv支不支持, 不支持你可以自己写一个, 我传的大概会是这样子:
```
mpv://bilibili?url=<direct url>
```

**Reply by Lee-7723** [reply]
那我可有不少要学的了2333

**Reply by Lee-7723** [reply]
这种格式吧
`mpv-bilibili://bilibili?url=<direct url>`

例如
`mpv-bilibili://bilibili?url=https://cn-tj2-cmcc-bcache-08.bilivideo.com/upgcxcode/03/08/327030803/327030803-1-116.flv?e=ig8euxZM2rNcNbRjhbdVhwdlhWTghwdVhoNvNC8BqJIzNbfqXBvEqxTEto8BTrNvN0GvT90W5JZMkX_YN0MvXg8gNEV4NC8xNEV4N03eN0B5tZlqNxTEto8BTrNvNeZVuJ10Kj_g2UB02J0mN0B5tZlqNCNEto8BTrNvNC7MTX502C8f2jmMQJ6mqF2fka1mqx6gqj0eN0B599M=&uipk=5&nbs=1&deadline=1619086448&gen=playurlv2&os=bcache&oi=3746063178&trid=80dcfe21fbe3435ca6599c15a2c46199u&platform=pc&upsig=22f25e1889ee88787785300e825b5545&uparams=e,uipk,nbs,deadline,gen,os,oi,trid,platform&cdnid=6779&mid=310265481&orderid=0,3&agrr=1&logo=80000000`

**Reply by Lee-7723** [reply]
没写过bat，就简单了解到让上面那个链接能跑起来而已😂

**Reply by the1812** [reply]
还有个问题, 有多个链接的时候咋办, 有些老视频是分段的

**Reply by Lee-7723** [reply]
完蛋，没考虑，滚回去继续学习😥

**Reply by Lee-7723** [reply]
所以基本解决思路是类似 #1472 这种拼接的办法（mpv可能不支持concat，不过可以生成一个m3u8列表文件，也能凑合接上）
要么是用m4s（所有m4s都不分段吧？），我去翻翻mpv能不能音频视频分开输入

**Reply by the1812** [reply]
或者干脆就不考虑老视频了? 遇到的几率应该也不大

**Reply by Lee-7723** [reply]
“先挖个坑”

**Reply by Lee-7723** [reply]
那您有时间的时候就先做一个出来吧，后续再完善
阿里云盘做这么个功能有用的多，网页端的下载直链加referer就可以取得文件，试过了，和B站这个同理，大佬有兴趣做个脚本么（手动滑稽）

**Reply by the1812** [reply]
我在下一个版本会开放下载视频的导出扩展, 链接啥的都会传过来, 你自由发挥就行了
<img width="400" src="https://user-images.githubusercontent.com/26504152/115701916-e360e000-a39a-11eb-8d31-63c86377cdbd.png">

**Reply by Lee-7723** [reply]
我来填坑了，mpv可以指定单独音频轨道，这样的话就可以用音频和视频的m4s直链播放了。
而且可以多个输入文件之间切换视频/音频轨道，这样的话应该都可以在mpv中选择分辨率了。
我先去看看怎么从url传递过来

**Reply by Lee-7723** [reply]
虽然写的完全不对，但是可以运行。。。
url格式是：
`mpv-bilibili://bilibili?url=<direct m4s video url>@<direct m4s audio url>`
例如：
`mpv-bilibili://bilibili?url=https://cn-tj2-cmcc-bcache-07.bilivideo.com/upgcxcode/69/19/327441969/327441969-1-30116.m4s?e=ig8euxZM2rNcNbdlhoNvNC8BqJIzNbfqXBvEqxTEto8BTrNvN0GvT90W5JZMkX_YN0MvXg8gNEV4NC8xNEV4N03eN0B5tZlqNxTEto8BTrNvNeZVuJ10Kj_g2UB02J0mN0B5tZlqNCNEto8BTrNvNC7MTX502C8f2jmMQJ6mqF2fka1mqx6gqj0eN0B599M=&uipk=5&nbs=1&deadline=1619103211&gen=playurlv2&os=bcache&oi=3746062882&trid=1f051c5903aa402bba9997c96340722au&platform=pc&upsig=d76d0b489fc75e329f0513ded489a50d&uparams=e,uipk,nbs,deadline,gen,os,oi,trid,platform&cdnid=6778&mid=310265481&orderid=0,3&agrr=0&logo=80000000@https://cn-hbsjz-cmcc-bcache-07.bilivideo.com/upgcxcode/69/19/327441969/327441969-1-30280.m4s?e=ig8euxZM2rNcNbdlhoNvNC8BqJIzNbfqXBvEqxTEto8BTrNvN0GvT90W5JZMkX_YN0MvXg8gNEV4NC8xNEV4N03eN0B5tZlqNxTEto8BTrNvNeZVuJ10Kj_g2UB02J0mN0B5tZlqNCNEto8BTrNvNC7MTX502C8f2jmMQJ6mqF2fka1mqx6gqj0eN0B599M=&uipk=5&nbs=1&deadline=1619103211&gen=playurlv2&os=bcache&oi=3746062882&trid=1f051c5903aa402bba9997c96340722au&platform=pc&upsig=b4ca277eca1e0484d70bebc15b60bde7&uparams=e,uipk,nbs,deadline,gen,os,oi,trid,platform&cdnid=5212&mid=310265481&orderid=0,3&agrr=0&logo=80000000`

**Reply by Lee-7723** [reply]
测试的时候遇到了少数几次400 bad request，原因未知

**Reply by Lee-7723** [reply]
这样的话会通用一些，简单来说就是我以@作为两个变量的分界，@之前的作为主输入文件，@之后的作为挂载音轨，应该能通吃m4s和flv（flv就是会报个错说挂载音轨的地方是空的，但还是会播放的）
`mpv-bilibili://bilibili?url=<direct m4s video url>@<direct m4s audio url>`
`mpv-bilibili://bilibili?url=<direct flv video url>@`

**Reply by sz1262011** [reply]
![image](https://user-images.githubusercontent.com/13992122/115957452-2656bb00-a535-11eb-8a4f-b22fe230c819.png)
@Lijingze2019  mpv 有参数可以接收两个流,多个流要加--merge-files. 我在知乎后文第一个参考链接如下:
https://fspark.me/archives/Linux-mpv-bilibili-bangumi-you-get.html#toc-2you-get%E8%8E%B7%E5%8F%96%E8%A7%A3%E6%9E%90%E5%9C%B0%E5%9D%80

![image](https://user-images.githubusercontent.com/13992122/115957495-69189300-a535-11eb-96cf-ad30b74f8649.png)

**Reply by Lee-7723** [reply]
谢谢，翻mpv文档的时候翻到了external-files audio-files这两种解决方案，没注意到merge-files这个，谢谢提醒。不过我不是很会写bat，现在能做到，<direct url>给完整flv的链接可以，给音视频分开的两个m4s链接也可以，但是分段flv的话给进来就要写个判断函数了，不太会。

**Reply by sz1262011** [reply]
> 谢谢，翻mpv文档的时候翻到了external-files audio-files这两种解决方案，没注意到merge-files这个，谢谢提醒。不过我不是很会写bat，现在能做到，给完整flv的链接可以，给音视频分开的两个m4s链接也可以，但是分段flv的话给进来就要写个判断函数了，不太会。

@Lijingze2019 

flv分集:
you-get -p "mpv --referrer='https://www.bilibili.com' --merge-files" https://www.bilibili.com/video/BV1ah41127s7 --playlist

m4s分集:
you-get -p "mpv --merge-files" http://www.acfun.cn/v/ac3262317

m3u8线上流拉取:
you-get -u https://www.acfun.cn/v/ac3262317 -playlists

**Reply by Lee-7723** [reply]
想绕过youtube-dl或者you-get，主要是解析速度慢，传递直链的话基本就秒开了。对于多P视频，就先用网页作为导航吧。（其实主要是我不会写那种bat，而且1812要改的应该也会多一些）

**Reply by sz1262011** [reply]
https://github.com/mpv-player/mpv/issues/1178
https://stackoverflow.com/questions/49343174/can-ffmpeg-concatenate-files-from-a-different-domain 
参考如上

**Comment by Lee-7723** [comment]
好耶！

**Comment by the1812** [comment]
按照你上面的格式, 可以这么弄一个脚本: (不过我没 mpv 所以没测过能不能运行)
```js
// ==UserScript==
// @name          mpv 播放
// @version       1.0.0
// @author        Grant Howard
// @match         https://www.bilibili.com/video/*
// @grant         unsafeWindow
// ==/UserScript==

;(async () => {
  if (!unsafeWindow.bilibiliEvolved) {
    unsafeWindow.bilibiliEvolved = {}
  }
  if (!unsafeWindow.bilibiliEvolved.downloadVideoExtensionButtons) {
    unsafeWindow.bilibiliEvolved.downloadVideoExtensionButtons = []
  }
  unsafeWindow.bilibiliEvolved.downloadVideoExtensionButtons.push({
    name: 'mpv',
    displayName: 'mpv 播放',
    batch: false,
    action: video => {
      const urls = video.fragments.map(it => it.url).join('@')
      const link = `mpv-bilibili://bilibili?url=${urls}`
      window.open(link)
    }
  })
})()
```
文档: https://github.com/the1812/Bilibili-Evolved/blob/preview/doc/download-video-extensions.md

**Reply by Lee-7723** [reply]
**PERFECT！**
https://user-images.githubusercontent.com/52600662/117526796-18f00500-affa-11eb-9787-093ec7b04897.mp4

**Reply by Lee-7723** [reply]
![image](https://user-images.githubusercontent.com/52600662/117527004-3c677f80-affb-11eb-8da5-a130b8dfe127.png)
m4s和flv都没太大问题

**Reply by Lee-7723** [reply]
专栏：https://www.bilibili.com/read/cv11201552
